File: test/org/stringtemplate/test/TestOptions.java
package org.stringtemplate.test;

import org.junit.Test;
import static org.junit.Assert.assertEquals;
import org.stringtemplate.ST;
import org.stringtemplate.STGroup;

import java.util.List;
import java.util.ArrayList;

public class TestOptions extends BaseTest {
    @Test public void testSeparator() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "hi <name; separator=\", \">!");
        ST st = group.getInstanceOf("test");
        st.add("name", "Ter");
        st.add("name", "Tom");
        st.add("name", "Sumana");
        String expected = "hi Ter, Tom, Sumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testAttrSeparator() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "hi <name; separator=sep>!");
        ST st = group.getInstanceOf("test");
        st.add("sep", ", ");
        st.add("name", "Ter");
        st.add("name", "Tom");
        st.add("name", "Sumana");
        String expected = "hi Ter, Tom, Sumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testIncludeSeparator() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("foo", "|");
        group.defineTemplate("test", "hi <name; separator=foo()>!");
        ST st = group.getInstanceOf("test");
        st.add("sep", ", ");
        st.add("name", "Ter");
        st.add("name", "Tom");
        st.add("name", "Sumana");
        String expected = "hi Ter|Tom|Sumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testSubtemplateSeparator() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "hi <name; separator={<sep> _}>!");
        ST st = group.getInstanceOf("test");
        st.add("sep", ",");
        st.add("name", "Ter");
        st.add("name", "Tom");
        st.add("name", "Sumana");
        String expected = "hi Ter, _Tom, _Sumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testSeparatorWithNullFirstValueAndNullOption() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "hi <name; null=\"n/a\", separator=\", \">!");
        ST st = group.getInstanceOf("test");
        st.add("name", null);
        st.add("name", "Tom");
        st.add("name", "Sumana");
        String expected = "hi n/a, Tom, Sumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testSeparatorWithNull2ndValueAndNullOption() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "hi <name; null=\"n/a\", separator=\", \">!");
        ST st = group.getInstanceOf("test");
        st.add("name", "Ter");
        st.add("name", null);
        st.add("name", "Sumana");
        String expected = "hi Ter, n/a, Sumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testNullValueAndNullOption() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "<name; null=\"n/a\">");
        ST st = group.getInstanceOf("test");
        st.add("name", null);
        String expected = "n/a";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testMissingValueAndNullOption() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "<name; null=\"n/a\">");
        ST st = group.getInstanceOf("test");
        String expected = "n/a";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testOptionDoesntApplyToNestedTemplate() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("foo", "<zippo>");
        group.defineTemplate("test", "<foo(); null=\"n/a\">");
        ST st = group.getInstanceOf("test");
        st.add("zippo", null);
        String expected = "";
        String result = st.render();
        assertEquals(expected, result);
    }
}

File: objc/main/tests/output/TestOptions.m
#import "TestOptions.h"

@implementation TestOptions

- (void) testSeparator {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"name" param2:@"hi <name; separator=\", \">!"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"name" param1:@"Ter"];
  [st add:@"name" param1:@"Tom"];
  [st add:@"name" param1:@"Sumana"];
  NSString * expected = @"hi Ter, Tom, Sumana!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testSeparatorWithSpaces {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  STGroup.debug = YES;
  [group defineTemplate:@"test" param1:@"name" param2:@"hi <name; separator= \", \">!"];
  ST * st = [group getInstanceOf:@"test"];
  [System.out println:[st.impl.ast toStringTree]];
  [st add:@"name" param1:@"Ter"];
  [st add:@"name" param1:@"Tom"];
  [st add:@"name" param1:@"Sumana"];
  NSString * expected = @"hi Ter, Tom, Sumana!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testAttrSeparator {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"name,sep" param2:@"hi <name; separator=sep>!"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"sep" param1:@", "];
  [st add:@"name" param1:@"Ter"];
  [st add:@"name" param1:@"Tom"];
  [st add:@"name" param1:@"Sumana"];
  NSString * expected = @"hi Ter, Tom, Sumana!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testIncludeSeparator {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"foo" param1:@"|"];
  [group defineTemplate:@"test" param1:@"name,sep" param2:@"hi <name; separator=foo()>!"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"sep" param1:@", "];
  [st add:@"name" param1:@"Ter"];
  [st add:@"name" param1:@"Tom"];
  [st add:@"name" param1:@"Sumana"];
  NSString * expected = @"hi Ter|Tom|Sumana!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testSubtemplateSeparator {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"name,sep" param2:@"hi <name; separator={<sep> _}>!"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"sep" param1:@","];
  [st add:@"name" param1:@"Ter"];
  [st add:@"name" param1:@"Tom"];
  [st add:@"name" param1:@"Sumana"];
  NSString * expected = @"hi Ter, _Tom, _Sumana!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testSeparatorWithNullFirstValueAndNullOption {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"name" param2:@"hi <name; null=\"n/a\", separator=\", \">!"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"name" param1:nil];
  [st add:@"name" param1:@"Tom"];
  [st add:@"name" param1:@"Sumana"];
  NSString * expected = @"hi n/a, Tom, Sumana!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testSeparatorWithNull2ndValueAndNullOption {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"name" param2:@"hi <name; null=\"n/a\", separator=\", \">!"];
  ST * st = [group getInstanceOf:@"test"];
  [st.impl dump];
  [st add:@"name" param1:@"Ter"];
  [st add:@"name" param1:nil];
  [st add:@"name" param1:@"Sumana"];
  NSString * expected = @"hi Ter, n/a, Sumana!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testNullValueAndNullOption {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"name" param2:@"<name; null=\"n/a\">"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"name" param1:nil];
  NSString * expected = @"n/a";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testListApplyWithNullValueAndNullOption {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"name" param2:@"<name:{n | <n>}; null=\"n/a\">"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"name" param1:@"Ter"];
  [st add:@"name" param1:nil];
  [st add:@"name" param1:@"Sumana"];
  NSString * expected = @"Tern/aSumana";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testDoubleListApplyWithNullValueAndNullOption {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"name" param2:@"<name:{n | <n>}:{n | [<n>]}; null=\"n/a\">"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"name" param1:@"Ter"];
  [st add:@"name" param1:nil];
  [st add:@"name" param1:@"Sumana"];
  NSString * expected = @"[Ter]n/a[Sumana]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testMissingValueAndNullOption {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"name" param2:@"<name; null=\"n/a\">"];
  ST * st = [group getInstanceOf:@"test"];
  NSString * expected = @"n/a";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testOptionDoesntApplyToNestedTemplate {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"foo" param1:@"<zippo>"];
  [group defineTemplate:@"test" param1:@"zippo" param2:@"<foo(); null=\"n/a\">"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"zippo" param1:nil];
  NSString * expected = @"";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testIllegalOption {
  ErrorBuffer * errors = [[[ErrorBuffer alloc] init] autorelease];
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group setListener:errors];
  [group defineTemplate:@"test" param1:@"name" param2:@"<name; bad=\"ugly\">"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"name" param1:@"Ter"];
  NSString * expected = @"Ter";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
  expected = @"[test 1:7: no such option: bad]";
  [self assertEquals:expected param1:[errors.errors description]];
}

@end
